File: frontend/reddit-frontend-starter/src/services/communityService.js
import api from "./api";

/* =========================================
   GET ALL COMMUNITIES
========================================= */

export const getAllCommunities =
  async () => {

    try {

      const response =
        await api.get(
          "/api/communities"
        );

      /* SAFE ARRAY */

      const communities =

        Array.isArray(response.data)

          ? response.data

          : [];

      return communities.map(
        (community) => ({

          ...community,

          id:
            community.id || null,

          name:
            community.name ||
            "Unknown Community",

          description:
            community.description || "",

          posts:
            community.posts || []

        })
      );

    } catch (error) {

      console.log(
        "GET COMMUNITIES ERROR :",
        error.response?.data || error.message
      );

      return [];

    }

  };

/* =========================================
   GET COMMUNITY BY ID
========================================= */

export const getCommunityById =
  async (communityId) => {

    try {

      const response =
        await api.get(

          `/api/communities/${communityId}`

        );

      return response.data;

    } catch (error) {

      console.log(
        "GET COMMUNITY ERROR :",
        error.response?.data || error.message
      );

      return null;

    }

  };

/* =========================================
   CREATE COMMUNITY
========================================= */

export const createCommunity =
  async (communityData) => {

    try {

      /* SAFE PAYLOAD */

      const payload = {

        name:
          communityData.name?.trim(),

        description:
          communityData.description?.trim()

      };

      const response =
        await api.post(

          "/api/communities",

          payload

        );

      return response.data;

    } catch (error) {

      console.log(
        "CREATE COMMUNITY ERROR :",
        error.response?.data || error.message
      );

      throw error;

    }

  };

/* =========================================
   GET POSTS BY COMMUNITY
========================================= */

export const getPostsByCommunity =
  async (communityId) => {

    try {

      const response =
        await api.get(

          `/api/posts/community/${communityId}`

        );

      const posts =

        Array.isArray(response.data)

          ? response.data

          : [];

      /* SAFE FORMAT */

      return posts.map((post) => ({

        ...post,

        communityName:

          post.community?.name ||

          "General",

        communityId:

          post.community?.id ||

          null,

        author:

          post.author ||

          "Anonymous",

        votes:

          post.votes || 0,

        comments:

          post.comments || 0,

        imageUrl:

          post.imageUrl || ""

      }));

    } catch (error) {

      console.log(
        "GET POSTS ERROR :",
        error.response?.data || error.message
      );

      return [];

    }

  };

/* =========================================
   JOIN COMMUNITY
========================================= */

export const joinCommunity =
  async (

    communityId,

    username

  ) => {

    try {

      const response =
        await api.post(

          `/api/communities/${communityId}/join`,

          null,

          {

            params: {
              username
            }

          }

        );

      return response.data;

    } catch (error) {

      console.log(
        "JOIN COMMUNITY ERROR :",
        error.response?.data || error.message
      );

      throw error;

    }

  };

/* =========================================
   LEAVE COMMUNITY
========================================= */

export const leaveCommunity =
  async (

    communityId,

    username

  ) => {

    try {

      const response =
        await api.delete(

          `/api/communities/${communityId}/leave`,

          {

            params: {
              username
            }

          }

        );

      return response.data;

    } catch (error) {

      console.log(
        "LEAVE COMMUNITY ERROR :",
        error.response?.data || error.message
      );

      throw error;

    }

  };

/* =========================================
   GET MEMBER COUNT
========================================= */

export const getMemberCount =
  async (communityId) => {

    try {

      const response =
        await api.get(

          `/api/communities/${communityId}/members`

        );

      return (
        response.data?.members || 0
      );

    } catch (error) {

      console.log(
        "MEMBER COUNT ERROR :",
        error.response?.data || error.message
      );

      return 0;

    }

  };

/* =========================================
   CHECK JOINED STATUS
========================================= */

export const isJoined =
  async (

    communityId,

    username

  ) => {

    try {

      const response =
        await api.get(

          `/api/communities/${communityId}/joined`,

          {

            params: {
              username
            }

          }

        );

      return (
        response.data?.joined || false
      );

    } catch (error) {

      console.log(
        "JOINED STATUS ERROR :",
        error.response?.data || error.message
      );

      return false;

    }

  };

File: backend/reddit-backend-starter/src/main/java/com/socialmedia/backend/service/CommunityService.java
package com.socialmedia.backend.service;

import com.socialmedia.backend.entity.Community;
import com.socialmedia.backend.entity.CommunityMember;
import com.socialmedia.backend.repository.CommunityMemberRepository;
import com.socialmedia.backend.repository.CommunityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class CommunityService {

    @Autowired
    private CommunityRepository
            communityRepository;

    @Autowired
    private CommunityMemberRepository
            communityMemberRepository;

    /* Create Community */

    public Community createCommunity(
            Community community
    ) {

        /* Null Validation */

        if (community == null) {

            throw new RuntimeException(
                    "Community data is missing"
            );

        }

        /* Name Validation */

        if (
                community.getName() == null ||
                        community.getName()
                                .trim()
                                .isEmpty()
        ) {

            throw new RuntimeException(
                    "Community name is required"
            );

        }

        /* Trim Name */

        community.setName(
                community.getName()
                        .trim()
        );

        /* Trim Description */

        if (
                community.getDescription() != null
        ) {

            community.setDescription(
                    community.getDescription()
                            .trim()
            );

        }

        /* Duplicate Check */

        boolean exists =
                communityRepository
                        .findByNameIgnoreCase(
                                community.getName()
                        )
                        .isPresent();

        if (exists) {

            throw new RuntimeException(
                    "Community already exists"
            );

        }

        /* Created Time */

        if (
                community.getCreatedAt() == null
        ) {

            community.setCreatedAt(
                    LocalDateTime.now()
            );

        }

        /* Save */

        return communityRepository
                .save(community);

    }

    /* Get All Communities */

    public List<Community>
    getAllCommunities() {

        return communityRepository
                .findAll();

    }

    /* Get Community By Id */

    public Community getCommunityById(
            Long id
    ) {

        return communityRepository
                .findById(id)

                .orElseThrow(

                        () -> new RuntimeException(
                                "Community Not Found"
                        )

                );

    }

    /* Join Community */

    public String joinCommunity(
            Long communityId,
            String username
    ) {

        Community community =
                getCommunityById(
                        communityId
                );

        boolean alreadyJoined =

                communityMemberRepository
                        .findByUsernameAndCommunity(
                                username,
                                community
                        )
                        .isPresent();

        if (alreadyJoined) {

            return "Already Joined";

        }

        CommunityMember member =
                new CommunityMember();

        member.setUsername(
                username
        );

        member.setCommunity(
                community
        );

        communityMemberRepository
                .save(member);

        return "Community Joined";

    }

    /* Leave Community */

    public String leaveCommunity(
            Long communityId,
            String username
    ) {

        Community community =
                getCommunityById(
                        communityId
                );

        CommunityMember member =

                communityMemberRepository
                        .findByUsernameAndCommunity(
                                username,
                                community
                        )

                        .orElseThrow(

                                () -> new RuntimeException(
                                        "You are not a member"
                                )

                        );

        communityMemberRepository
                .delete(member);

        return "Community Left";

    }

    /* Member Count */

    public long getMemberCount(
            Long communityId
    ) {

        Community community =
                getCommunityById(
                        communityId
                );

        return communityMemberRepository
                .countByCommunity(
                        community
                );

    }

    /* Joined Status */

    public boolean isJoined(
            Long communityId,
            String username
    ) {

        Community community =
                getCommunityById(
                        communityId
                );

        return communityMemberRepository
                .findByUsernameAndCommunity(
                        username,
                        community
                )
                .isPresent();

    }

}